File: frontend/lib/recommender.ts
import type { FinancialData, FinancialRatios, Recommendation } from "./types";
import { derivePL, deriveBS } from "./calculator";

export function generateRecommendations(
  data: FinancialData,
  ratios: FinancialRatios,
  prevRatios?: FinancialRatios | null
): Recommendation[] {
  const recs: Recommendation[] = [];
  const bs = data.balance_sheet;
  const pl = data.profit_loss;
  const p = derivePL(pl);
  const d = deriveBS(bs);
  const rev = pl.revenue_from_operations;

  // ── HIGH PRIORITY ──────────────────────────────────────────────
  if (bs.pf_esi_payable > 0) {
    recs.push({
      priority: "HIGH",
      category: "Compliance",
      title: `PF/ESI Payable Outstanding — ₹${bs.pf_esi_payable.toFixed(1)}L`,
      description:
        "Directors face personal criminal liability under EPF/ESIC Acts for payment defaults. This is non-negotiable — deposit immediately via EPFO/ESIC portal.",
      impact: "Avoid criminal prosecution and compounding interest",
      action: "Deposit via EPFO portal before next salary cycle",
    });
  }

  if (bs.tds_payable > 0) {
    recs.push({
      priority: "HIGH",
      category: "Compliance",
      title: `TDS Payable Outstanding — ₹${bs.tds_payable.toFixed(1)}L`,
      description:
        "TDS deducted but not deposited attracts 1.5% interest per month under Section 201 of the Income Tax Act, plus potential prosecution of TAN holders.",
      impact: "Stop 1.5%/month interest accrual",
      action: "Deposit via NSDL challan ITNS 281 before 7th of next month",
    });
  }

  if (ratios.dscr !== undefined && ratios.dscr < 1.0) {
    recs.push({
      priority: "HIGH",
      category: "Debt",
      title: `DSCR at ${ratios.dscr.toFixed(2)}x — NPA Risk`,
      description:
        "DSCR below 1.0x means operating cash flow cannot cover debt obligations. Under RBI classification, your account may be heading toward NPA status. Seek loan restructuring immediately.",
      impact: "Prevent NPA classification and facility recall",
      action: "Approach your bank for restructuring or moratorium under RBI MSME schemes",
    });
  } else if (ratios.dscr !== undefined && ratios.dscr < 1.25) {
    recs.push({
      priority: "HIGH",
      category: "Debt",
      title: `DSCR Below RBI Minimum — ${ratios.dscr.toFixed(2)}x`,
      description:
        "RBI requires DSCR ≥ 1.25x for Standard Asset classification. Your current DSCR puts the account in Sub-Standard territory. A 90-day overdue will trigger NPA.",
      impact: "Prevent NPA classification",
      action: "Accelerate collections; defer non-critical expenditure",
    });
  }

  if (ratios.working_capital !== undefined && ratios.working_capital < 0) {
    recs.push({
      priority: "HIGH",
      category: "Liquidity",
      title: `Negative Working Capital — ₹${Math.abs(ratios.working_capital).toFixed(1)}L Shortfall`,
      description:
        "Current liabilities exceed current assets. The company technically cannot meet short-term obligations from current assets alone. Payment defaults are imminent.",
      impact: "Prevent vendor payment defaults and bounced cheques",
      action: "Convert long-term debt to short-term; accelerate debtor collections immediately",
    });
  }

  if (ratios.current_ratio !== undefined && ratios.current_ratio < 1.0) {
    recs.push({
      priority: "HIGH",
      category: "Liquidity",
      title: `Critical Current Ratio — ${ratios.current_ratio.toFixed(2)}x`,
      description:
        "Current ratio below 1.0x indicates a short-term solvency crisis. For every ₹1 of current dues, you have less than ₹1 in current assets.",
      impact: "Prevent immediate payment defaults",
      action: "Negotiate extended credit from suppliers; bring in short-term equity",
    });
  }

  if (ratios.debt_to_equity !== undefined && ratios.debt_to_equity > 3.0) {
    recs.push({
      priority: "HIGH",
      category: "Debt",
      title: `IBC Risk — D/E Ratio at ${ratios.debt_to_equity.toFixed(2)}x`,
      description:
        "If any creditor has a default of ₹1 Crore or more, they can initiate Corporate Insolvency Resolution Process (CIRP) under the Insolvency & Bankruptcy Code. D/E above 3x is a serious red flag.",
      impact: "Prevent loss of management control under IBC",
      action: "Raise equity capital or convert promoter loans to equity immediately",
    });
  }

  if (ratios.dso !== undefined && ratios.dso > 120) {
    recs.push({
      priority: "HIGH",
      category: "Receivables",
      title: `Very High DSO — ${ratios.dso.toFixed(0)} Days`,
      description:
        "Debtors outstanding for more than 120 days are likely to become bad debts. This also means 4+ months of revenue is trapped with customers.",
      impact: "Release working capital and reduce bad debt write-offs",
      action: "Issue legal notices to 90+ day overdue customers; engage collection agency; consider invoice discounting",
    });
  }

  if (rev > 0 && pl.finance_costs / rev > 0.08) {
    recs.push({
      priority: "HIGH",
      category: "Debt",
      title: `Finance Costs Very High — ${((pl.finance_costs / rev) * 100).toFixed(1)}% of Revenue`,
      description:
        `Finance costs consuming ${((pl.finance_costs / rev) * 100).toFixed(1)}% of revenue (healthy: <5%). This severely compresses net margins.`,
      impact: "Improve net margin by 2–4 percentage points",
      action: "Refinance high-cost loans; prepay CC/OD with surplus cash; explore MSME Samadhaan loan schemes",
    });
  }

  // ── MEDIUM PRIORITY ────────────────────────────────────────────
  if (ratios.dso !== undefined && ratios.dso > 60 && ratios.dso <= 120) {
    recs.push({
      priority: "MEDIUM",
      category: "Receivables",
      title: `DSO at ${ratios.dso.toFixed(0)} Days — Above Ideal`,
      description:
        "Industry ideal for Indian SMEs is <60 days. Introduce early payment discounts (2/15 net 45) to incentivise faster payments and reduce CC/OD utilisation.",
      impact: "Release working capital; reduce interest on CC",
      action: "Offer 1–2% discount for payment within 15 days",
    });
  }

  if (ratios.ccc !== undefined && ratios.ccc > 90) {
    recs.push({
      priority: "MEDIUM",
      category: "Efficiency",
      title: `Cash Conversion Cycle Too Long — ${ratios.ccc.toFixed(0)} Days`,
      description:
        `CCC of ${ratios.ccc.toFixed(0)} days means it takes ${(ratios.ccc / 30).toFixed(1)} months to convert inventory investment back into cash. High working capital requirement increases debt.`,
      impact: "Reduce CC/OD dependency",
      action: "Reduce inventory holding; negotiate better supplier payment terms (longer DPO)",
    });
  }

  if (ratios.dio !== undefined && ratios.dio > 90) {
    recs.push({
      priority: "MEDIUM",
      category: "Efficiency",
      title: `High Inventory Days — ${ratios.dio.toFixed(0)} Days`,
      description:
        "Inventory held for more than 3 months before sale indicates overstock or slow-moving products. This locks working capital and increases storage and obsolescence costs.",
      impact: "Release cash locked in inventory",
      action: "Identify slow-moving SKUs; implement JIT ordering; run clearance sales",
    });
  }

  if (rev > 0 && pl.employee_expenses / rev > 0.3) {
    recs.push({
      priority: "MEDIUM",
      category: "Expenses",
      title: `Labour Costs High — ${((pl.employee_expenses / rev) * 100).toFixed(1)}% of Revenue`,
      description:
        `Labour costs at ${((pl.employee_expenses / rev) * 100).toFixed(1)}% of revenue (threshold: 30%). Review staffing productivity and explore automation.`,
      impact: "Improve EBITDA margin by 3–5 percentage points",
      action: "Productivity analysis; explore automation ROI; review contract workforce",
    });
  }

  if (rev > 0 && bs.gst_itc_receivable > (rev / 12) * 3) {
    recs.push({
      priority: "MEDIUM",
      category: "Compliance",
      title: `GST ITC Blocked — ${bs.gst_itc_receivable.toFixed(1)}L`,
      description: `More than 3 months of GST Input Tax Credit is blocked (₹${bs.gst_itc_receivable.toFixed(1)}L). This is locked working capital with the government.`,
      impact: `Release ₹${bs.gst_itc_receivable.toFixed(1)}L of cash`,
      action: "Reconcile GSTR-2A vs purchase register; utilise ITC against pending GST liability; file GSTR-3B accurately",
    });
  }

  if (prevRatios && ratios.net_margin !== undefined && prevRatios.net_margin !== undefined) {
    const drop = prevRatios.net_margin - ratios.net_margin;
    if (drop > 1.5) {
      recs.push({
        priority: "MEDIUM",
        category: "Profitability",
        title: `Net Margin Declining — Down ${drop.toFixed(1)}pp`,
        description: `Net margin fell from ${prevRatios.net_margin.toFixed(1)}% to ${ratios.net_margin.toFixed(1)}%. Review whether this is input cost pressure, labour, or finance cost driven.`,
        impact: "Restore profitability",
        action: "Cost reduction task force; review pricing for key products",
      });
    }
  }

  if (ratios.interest_coverage !== undefined && ratios.interest_coverage > 1.5 && ratios.interest_coverage < 3.0) {
    recs.push({
      priority: "MEDIUM",
      category: "Debt",
      title: `Interest Coverage Thin — ${ratios.interest_coverage.toFixed(2)}x`,
      description:
        "ICR between 1.5x and 3.0x is thin. A revenue dip of even 20–25% would make interest coverage dangerous.",
      impact: "Buffer against revenue volatility",
      action: "Consider fixed-rate refinancing to reduce interest rate risk",
    });
  }

  if (data.msme_payables > 0) {
    recs.push({
      priority: "MEDIUM",
      category: "Compliance",
      title: `MSME Payables — ₹${data.msme_payables.toFixed(1)}L`,
      description:
        "Under MSME Act Section 15, payment to MSME suppliers must be made within 45 days of acceptance. Overdue attracts compound interest at 3× bank rate. Disclosure mandatory in MCA filing.",
      impact: "Avoid mandatory interest payment to MSME suppliers",
      action: "Prioritise MSME supplier payments; file Form MSME-1 if overdue",
    });
  }

  if (data.promoter_loans > 0 && data.promoter_loans > (bs.long_term_borrowings + bs.short_term_borrowings) * 0.2) {
    recs.push({
      priority: "MEDIUM",
      category: "Debt",
      title: `High Promoter Loans — ₹${data.promoter_loans.toFixed(1)}L`,
      description:
        "Promoter/director loans exceeding 20% of total debt are scrutinised by banks and investors. Ensure proper documentation, board resolution, and RBI compliance for unsecured loans.",
      impact: "Improve creditworthiness and governance",
      action: "Convert promoter loans to equity; ensure proper documentation under Companies Act Sec 185/186",
    });
  }

  // ── POSITIVE REINFORCEMENT ─────────────────────────────────────
  if (ratios.debt_to_equity !== undefined && ratios.debt_to_equity <= 0.5) {
    recs.push({
      priority: "POSITIVE",
      category: "Debt",
      title: `Excellent Leverage — D/E at ${ratios.debt_to_equity.toFixed(2)}x`,
      description:
        "D/E well below 1x shows strong financial discipline. The company has significant headroom to raise debt for growth if needed.",
      action: "Consider leveraged growth — project-linked debt with clear ROI",
    });
  }

  if (prevRatios?.debt_to_equity && ratios.debt_to_equity !== undefined) {
    const improvement = prevRatios.debt_to_equity - ratios.debt_to_equity;
    if (improvement > 0.2) {
      recs.push({
        priority: "POSITIVE",
        category: "Debt",
        title: `Deleveraging on Track — D/E Improved ${improvement.toFixed(2)}x`,
        description: `D/E improved from ${prevRatios.debt_to_equity.toFixed(2)}x to ${ratios.debt_to_equity.toFixed(2)}x. Debt reduction strategy is working well.`,
        action: "Continue debt repayment from operating cash flow",
      });
    }
  }

  if (ratios.current_ratio !== undefined && ratios.current_ratio >= 1.5 && ratios.current_ratio <= 2.5) {
    recs.push({
      priority: "POSITIVE",
      category: "Liquidity",
      title: `Healthy Liquidity — Current Ratio ${ratios.current_ratio.toFixed(2)}x`,
      description:
        "Current ratio in the ideal 1.5–2.5x range indicates good short-term solvency. The business can comfortably meet near-term obligations.",
      action: "Maintain working capital discipline; avoid over-stocking",
    });
  }

  if (ratios.dscr !== undefined && ratios.dscr >= 1.5) {
    recs.push({
      priority: "POSITIVE",
      category: "Debt",
      title: `Comfortable Debt Coverage — DSCR ${ratios.dscr.toFixed(2)}x`,
      description:
        `DSCR of ${ratios.dscr.toFixed(2)}x comfortably exceeds the RBI minimum of 1.25x. Loan accounts are in Standard Asset classification.`,
      action: "Maintain by protecting EBITDA margins and timely EMI payments",
    });
  }

  if (prevRatios?.revenue_from_operations && pl.revenue_from_operations > prevRatios.revenue_from_operations) {
    const growth = ((pl.revenue_from_operations - prevRatios.revenue_from_operations) / prevRatios.revenue_from_operations) * 100;
    if (growth > 10) {
      recs.push({
        priority: "POSITIVE",
        category: "Profitability",
        title: `Strong Revenue Growth — ${growth.toFixed(1)}%`,
        description: `Revenue grew ${growth.toFixed(1)}% year-over-year — strong top-line momentum. Focus on protecting margins while growing.`,
        action: "Now focus on operational leverage to improve margins alongside revenue growth",
      });
    }
  }

  // Sort: HIGH → MEDIUM → POSITIVE
  const order = { HIGH: 0, MEDIUM: 1, POSITIVE: 2 };
  return recs.sort((a, b) => order[a.priority] - order[b.priority]);
}

File: backend/services/recommender.py
"""
Recommendations engine — generates actionable insights from financial ratios.
Indian context with statutory compliance awareness.
"""
from typing import List, Optional
from models.financial_data import FinancialRatios, FinancialData, Recommendation


def generate_recommendations(
    ratios: FinancialRatios,
    data: FinancialData,
    prev_ratios: Optional[FinancialRatios] = None,
) -> List[Recommendation]:
    recommendations: List[Recommendation] = []
    bs = data.balance_sheet
    pl = data.profit_loss
    revenue = pl.revenue_from_operations

    # ── HIGH PRIORITY ──────────────────────────────────────────────

    # PF/ESI overdue — criminal liability risk
    if bs.pf_esi_payable > 0:
        recommendations.append(Recommendation(
            priority="HIGH",
            category="Compliance",
            title="PF/ESI Payable Outstanding",
            description=f"PF/ESI payable of ₹{bs.pf_esi_payable:.1f}L is outstanding. Directors face personal criminal liability under EPF Act for defaults. Deposit immediately.",
            impact="Avoid criminal prosecution and penalty interest",
            action="Deposit immediately via EPFO/ESIC portal",
        ))

    # TDS payable — interest + prosecution
    if bs.tds_payable > 0:
        recommendations.append(Recommendation(
            priority="HIGH",
            category="Compliance",
            title="TDS Payable Overdue",
            description=f"TDS payable of ₹{bs.tds_payable:.1f}L outstanding. Interest @1.5% per month under Section 201 of IT Act plus potential prosecution of TAN holders.",
            impact="Stop accumulating 1.5%/month interest",
            action="Deposit via NSDL before 7th of next month",
        ))

    # DSCR critical
    if ratios.dscr is not None and ratios.dscr < 1.0:
        recommendations.append(Recommendation(
            priority="HIGH",
            category="Debt",
            title="DSCR Below 1.0x — NPA Risk",
            description=f"DSCR of {ratios.dscr:.2f}x is below the RBI minimum of 1.25x. Loan may be classified as NPA. Operating cash flow cannot cover debt obligations.",
            impact="Prevent NPA classification and bank facility withdrawal",
            action="Negotiate loan restructuring or moratorium immediately",
        ))
    elif ratios.dscr is not None and ratios.dscr < 1.25:
        recommendations.append(Recommendation(
            priority="HIGH",
            category="Debt",
            title="DSCR Below RBI Minimum (1.25x)",
            description=f"DSCR of {ratios.dscr:.2f}x is in Sub-Standard zone (RBI norm: >1.25x). Risk of NPA classification within 90 days.",
            impact="Protect credit rating and bank relationships",
            action="Accelerate collections, defer non-critical capex",
        ))

    # Negative working capital
    if ratios.working_capital is not None and ratios.working_capital < 0:
        recommendations.append(Recommendation(
            priority="HIGH",
            category="Liquidity",
            title="Negative Working Capital",
            description=f"Working capital is negative (₹{abs(ratios.working_capital):.1f}L shortfall). Current liabilities exceed current assets — short-term solvency crisis.",
            impact="Prevent payment defaults to vendors and banks",
            action="Accelerate debtor collections, negotiate extended credit from suppliers",
        ))

    # Current ratio critical
    if ratios.current_ratio is not None and ratios.current_ratio < 1.0:
        recommendations.append(Recommendation(
            priority="HIGH",
            category="Liquidity",
            title="Critical Current Ratio — Below 1.0x",
            description=f"Current ratio of {ratios.current_ratio:.2f}x indicates inability to meet short-term obligations from current assets.",
            impact="Prevent payment defaults and creditor escalations",
            action="Convert long-term debt to short-term, accelerate receivables",
        ))

    # IBC risk — D/E > 3x
    if ratios.debt_to_equity is not None and ratios.debt_to_equity > 3.0:
        recommendations.append(Recommendation(
            priority="HIGH",
            category="Debt",
            title="IBC Risk — D/E Exceeds 3x",
            description=f"Debt-to-Equity of {ratios.debt_to_equity:.2f}x. If a creditor default exceeds ₹1 Cr, they can initiate CIRP under IBC. Seek equity infusion urgently.",
            impact="Prevent insolvency proceedings",
            action="Raise equity capital or convert promoter loans to equity",
        ))

    # Bad debt risk (high DSO + large receivables)
    if ratios.dso is not None and ratios.dso > 120:
        receivables_pct = (bs.trade_receivables / revenue * 100) if revenue else 0
        recommendations.append(Recommendation(
            priority="HIGH",
            category="Receivables",
            title=f"Very High DSO — {ratios.dso:.0f} Days",
            description=f"DSO of {ratios.dso:.0f} days with receivables at {receivables_pct:.1f}% of revenue indicates serious collection issues. Risk of bad debt write-offs.",
            impact="Free up working capital, reduce bad debt risk",
            action="Issue demand notices to 90+ day overdue customers; consider factoring",
        ))

    # Finance costs high
    if revenue > 0 and pl.finance_costs / revenue > 0.08:
        fc_pct = pl.finance_costs / revenue * 100
        recommendations.append(Recommendation(
            priority="HIGH",
            category="Debt",
            title=f"Finance Costs Very High ({fc_pct:.1f}% of Revenue)",
            description=f"Finance costs consuming {fc_pct:.1f}% of revenue (threshold: 5%). Severely compressing net margins.",
            impact="Improve net margin by reducing debt servicing burden",
            action="Refinance at lower rates, prepay high-cost loans with surplus cash",
        ))

    # ── MEDIUM PRIORITY ────────────────────────────────────────────

    # DSO moderate
    if ratios.dso is not None and 60 < ratios.dso <= 120:
        recommendations.append(Recommendation(
            priority="MEDIUM",
            category="Receivables",
            title=f"DSO at {ratios.dso:.0f} Days — Above Ideal",
            description=f"Industry ideal for Indian SMEs is <60 days. DSO of {ratios.dso:.0f} days indicates collections need improvement. Introduce early payment discounts.",
            impact="Release working capital, reduce CC/OD utilisation",
            action="Offer 1–2% discount for payment within 15 days (2/15, net 45)",
        ))

    # High CCC
    if ratios.ccc is not None and ratios.ccc > 90:
        recommendations.append(Recommendation(
            priority="MEDIUM",
            category="Efficiency",
            title=f"Cash Conversion Cycle Too Long ({ratios.ccc:.0f} Days)",
            description=f"CCC of {ratios.ccc:.0f} days means it takes over {ratios.ccc/30:.1f} months to convert inventory investment back to cash. High working capital requirement.",
            impact="Reduce working capital borrowing needs",
            action="Reduce inventory holding; negotiate better supplier terms",
        ))

    # High inventory days
    if ratios.dio is not None and ratios.dio > 90:
        recommendations.append(Recommendation(
            priority="MEDIUM",
            category="Efficiency",
            title=f"High Inventory Days ({ratios.dio:.0f} Days)",
            description=f"Inventory held for {ratios.dio:.0f} days before sale. Excess stock increases storage costs and obsolescence risk.",
            impact="Release cash locked in inventory",
            action="Implement JIT ordering, identify slow-moving SKUs for liquidation",
        ))

    # Employee costs high
    if revenue > 0 and pl.employee_expenses / revenue > 0.30:
        emp_pct = pl.employee_expenses / revenue * 100
        recommendations.append(Recommendation(
            priority="MEDIUM",
            category="Expenses",
            title=f"Employee Costs High ({emp_pct:.1f}% of Revenue)",
            description=f"Labour costs at {emp_pct:.1f}% of revenue (threshold: 30%). Review staffing efficiency or automate repetitive tasks.",
            impact="Improve EBITDA margin by 3–5 percentage points",
            action="Conduct productivity analysis; explore automation ROI",
        ))

    # GST ITC large
    if bs.gst_itc_receivable > 0 and revenue > 0:
        itc_months = bs.gst_itc_receivable / (revenue / 12) if revenue else 0
        if itc_months > 3:
            recommendations.append(Recommendation(
                priority="MEDIUM",
                category="Compliance",
                title=f"GST ITC Blocked — {itc_months:.1f} Months of Purchases",
                description=f"₹{bs.gst_itc_receivable:.1f}L of GST Input Tax Credit is blocked. More than 3 months of ITC not utilised. This is locked working capital.",
                impact="Release ₹{:.1f}L of cash blocked with government".format(bs.gst_itc_receivable),
                action="Reconcile GSTR-2A vs books; utilise ITC against GST liability",
            ))

    # Net margin declining
    if prev_ratios and ratios.net_margin is not None and prev_ratios.net_margin is not None:
        margin_decline = prev_ratios.net_margin - ratios.net_margin
        if margin_decline > 1.5:
            recommendations.append(Recommendation(
                priority="MEDIUM",
                category="Profitability",
                title=f"Net Margin Declining ({margin_decline:.1f}pp Drop)",
                description=f"Net margin fell from {prev_ratios.net_margin:.1f}% to {ratios.net_margin:.1f}%. Review cost structure — identify if it's input cost pressure, labour, or finance costs.",
                impact="Restore profitability to prior year levels",
                action="Cost reduction task force; price revision for key products",
            ))

    # Interest coverage moderate
    if ratios.interest_coverage is not None and 1.5 < ratios.interest_coverage < 3.0:
        recommendations.append(Recommendation(
            priority="MEDIUM",
            category="Debt",
            title=f"Interest Coverage Ratio Thin ({ratios.interest_coverage:.2f}x)",
            description=f"ICR of {ratios.interest_coverage:.2f}x is above minimum but thin. A revenue dip of >25% would make interest coverage dangerous.",
            impact="Build buffer against revenue volatility",
            action="Consider fixed-rate refinancing to reduce rate risk",
        ))

    # MSME payables
    if data.msme_payables > 0:
        recommendations.append(Recommendation(
            priority="MEDIUM",
            category="Compliance",
            title="MSME Payables — 45-Day Compliance",
            description=f"₹{data.msme_payables:.1f}L due to MSME suppliers. Under MSME Act Section 15, payment must be made within 45 days. Overdue attracts compound interest @3× bank rate.",
            impact="Avoid mandatory interest payment to MSME suppliers",
            action="Prioritise MSME supplier payments; disclose in MCA filing if overdue",
        ))

    # ── POSITIVE REINFORCEMENT ─────────────────────────────────────

    # Good D/E
    if ratios.debt_to_equity is not None and ratios.debt_to_equity < 0.5:
        recommendations.append(Recommendation(
            priority="POSITIVE",
            category="Debt",
            title="Excellent Leverage Position",
            description=f"D/E of {ratios.debt_to_equity:.2f}x is well below 1x. Company has strong capacity to raise debt for growth if needed.",
            action="Consider leveraged growth — take on project-linked debt with clear ROI",
        ))

    # Good D/E improving
    if prev_ratios and ratios.debt_to_equity and prev_ratios.debt_to_equity:
        if prev_ratios.debt_to_equity - ratios.debt_to_equity > 0.2:
            recommendations.append(Recommendation(
                priority="POSITIVE",
                category="Debt",
                title="Deleveraging on Track",
                description=f"D/E improved from {prev_ratios.debt_to_equity:.2f}x to {ratios.debt_to_equity:.2f}x. Debt reduction strategy is working.",
                action="Continue debt repayment from operating cash flow",
            ))

    # Good current ratio
    if ratios.current_ratio is not None and 1.5 <= ratios.current_ratio <= 2.5:
        recommendations.append(Recommendation(
            priority="POSITIVE",
            category="Liquidity",
            title="Healthy Liquidity Position",
            description=f"Current ratio of {ratios.current_ratio:.2f}x is in the ideal range (1.5–2.5x). Short-term solvency is strong.",
            action="Maintain working capital discipline; avoid over-stocking",
        ))

    # Revenue growth
    if prev_ratios:
        prev_rev = (prev_ratios.asset_turnover or 0) * 1000  # proxy
        # Use data directly
        pass

    # Good DSCR
    if ratios.dscr is not None and ratios.dscr >= 1.5:
        recommendations.append(Recommendation(
            priority="POSITIVE",
            category="Debt",
            title=f"Comfortable Debt Coverage (DSCR {ratios.dscr:.2f}x)",
            description=f"DSCR of {ratios.dscr:.2f}x comfortably exceeds the RBI minimum of 1.25x. Loan accounts are in Standard category.",
            action="Maintain by protecting EBITDA margins",
        ))

    # Sort: HIGH first, then MEDIUM, then POSITIVE
    priority_order = {"HIGH": 0, "MEDIUM": 1, "POSITIVE": 2}
    recommendations.sort(key=lambda r: priority_order.get(r.priority, 3))

    return recommendations
